cli/src/action/database.rs

```diff
@@ -12,37 +12,170 @@
+    sqlite::SqliteConnection,
+};
+
 #[cfg(feature = "database-migrate-biome")]
 use splinter::migrations::run_postgres_migrations;
+#[cfg(feature = "sqlite")]
+use splinter::migrations::run_sqlite_migrations;
+
+use crate::error::CliError;
+
+use super::Action;
+
+#[cfg(feature = "sqlite")]
+const SPLINTER_HOME_ENV: &str = "SPLINTER_HOME";
+#[cfg(feature = "sqlite")]
+const SPLINTER_STATE_DIR_ENV: &str = "SPLINTER_STATE_DIR";
+#[cfg(feature = "sqlite")]
+const DEFAULT_STATE_DIR: &str = "/var/lib/splinter";
+#[cfg(feature = "sqlite")]
+const DEFAULT_SQLITE: &str = "splinter_state.db";
 
 pub struct MigrateAction;
 
 impl Action for MigrateAction {
     fn run<'a>(&mut self, arg_matches: Option<&ArgMatches<'a>>) -> Result<(), CliError> {
         let url = if let Some(args) = arg_matches {
             args.value_of("connect")
-                .unwrap_or("postgres://admin:admin@localhost:5432/splinterd")
+                .map(ToOwned::to_owned)
+                .unwrap_or(get_default_database()?)
         } else {
-            "postgres://admin:admin@localhost:5432/splinterd"
+            get_default_database()?
         };
 
-        let connection = PgConnection::establish(url).map_err(|err| {
+        match ConnectionUri::from_str(&url).map_err(CliError::ActionError)? {
+            ConnectionUri::Postgres(url) => {
+                let connection = PgConnection::establish(&url).map_err(|err| {
+                    CliError::ActionError(format!(
+                        "Failed to establish database connection to '{}': {}",
+                        url, err
+                    ))
+                })?;
+
+                #[cfg(feature = "database-migrate-biome")]
+                run_postgres_migrations(&connection).map_err(|err| {
+                    CliError::ActionError(format!("Unable to run Biome migrations: {}", err))
+                })?;
+            }
+            #[cfg(feature = "sqlite")]
+            ConnectionUri::Sqlite(connection_string) => {
+                let connection_manager =
+                    ConnectionManager::<SqliteConnection>::new(&connection_string);
+                let mut pool_builder = Pool::builder();
+
+                if connection_string == ":memory:" {
+                    pool_builder = pool_builder.max_size(1)
+                }
+
+                let pool = pool_builder.build(connection_manager).map_err(|_| {
+                    CliError::ActionError("Failed to build connection pool".to_string())
+                })?;
+
+                run_sqlite_migrations(&*pool.get().map_err(|_| {
+                    CliError::ActionError("Failed to get connection for migrations".to_string())
+                })?)
+                .map_err(|err| {
+                    CliError::ActionError(format!("Unable to run Biome migrations: {}", err))
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn get_default_database() -> Result<String, CliError> {
+    #[cfg(feature = "sqlite")]
```

Comment:
Instead of doing this as blocks in the function, just create two functions with the guards here - it will be easier to follow what happens.